### DISCUSSION TITLE
v7 ref format is different from normal ref.current?

### DISCUSSION THREAD

**Discussion by hornetDC** [react-hook-form/react-hook-form#5680] [Q&A]
When I pass field.ref to the controlled component it looks like this:

```
(elm) => elm &&
                ref({
                    focus: () => elm.focus && elm.focus(),
                })
```

Which is not a usual react ref object with "current". I need to use the ref in the component to select html input for the google places.

`new window.google.maps.places.Autocomplete(inputRef.current, opts) `

**Comment by bluebill1049** [comment]
yes, for controller (controlled component) we only want the focus event, we don't really need the input reference (DOM) itself.

**Reply by hornetDC** [reply]
How can I get around it?

**Reply by bluebill1049** [reply]
what's the issue? not sure if i follow. can you share a csb?

**Reply by hornetDC** [reply]
I need a normal ref.current (React.RefObject) for google places to specify html input and RefCallback for focus. I can't use both refs. Before v7 Passing ref={register(...)} and using onFocus separately worked fine.

**Reply by hornetDC** [reply]
Rendering component (without field.ref)
```
            render={({ field: { ref, onChange, ...field } }) => (
              <InputPlaces
                onChange={onChange}
                onPlaceChange={(query, place) => {
                  handleAddressChange(place);
                  onChange(query);
                }}
                {...field}
              />
            )
```

Input component
```
const InputPlaces: React.FC<InputPlacesProps> = React.forwardRef(
  ({ options, onChange, onPlaceChange, value, ...props }, refProp) => {
    const ref = useRef<HTMLInputElement>(null);
    const inputRef = (refProp || ref) as React.RefObject<HTMLInputElement>;
    const autocompleteRef = useRef<google.maps.places.Autocomplete | null>(null);

    const handlePlaceSelect = useCallback(async () => {
      const autocomplete = autocompleteRef.current;
      if (!autocomplete) return;
      const place = autocomplete.getPlace();
      const query = place?.formatted_address || inputRef.current?.value;
      onPlaceChange?.(query, place);
    }, [inputRef, onPlaceChange]);

      if (autocompleteRef.current) return;

      const bounds = getBounds(BOUNDS_ILLINOIS);

      const opts: google.maps.places.AutocompleteOptions = {
        types: ['address'],
        componentRestrictions: { country: 'us' },
        bounds,
        ...options
      };
      autocompleteRef.current = new window.google.maps.places.Autocomplete(inputRef.current!, opts);
      autocompleteRef.current.setFields(['address_components', 'formatted_address', 'geometry']);
      autocompleteRef.current.addListener('place_changed', handlePlaceSelect);
    }, [handlePlaceSelect, inputRef, options]);

    return (
      <input
        ref={inputRef}
        defaultValue={value}
        onChange={onChange}
        {...props}
      />
    );
  }
);
```

EDIT:
I have been reusing forwardRef in a wrong way.
https://itnext.io/reusing-the-ref-from-forwardref-with-react-hooks-4ce9df693dd

**Reply by jcg-nw** [reply]
It looks like @hornetDC may have resolved his own issue, however I'm seeing the same problem where I've migrated some code from a project built in RHF v6, to one built in v7. We use a UI component library produced within our organisation that I don't have any control over unfortunately, and these components make use of ref.current in order to focus the field on label clicks. The problem I'm seeing is that now the v7 ref has changed from a ref object to a callback, I can't have both RHF auto-focusing fields when validation errors occur and working label clicks
